### PR TITLE
Add jnp.unravel_index

### LINKED ISSUE #2889: Add np.unravel_index
This is what I've been using. If it looks reasonable, will see about putting together a PR.

```python
def unravel_index(flat_index, shape):
    """
    To support abstract `flat_index` there is no bounds checking. For
    overflow some dimensions of the resulting index will be out of bounds.
    For negative `flat_index` the resulting index is valid, although the first
    dimension will be negative. 
    """

    if len(shape) == 0:
        return ()

    size_of_next_dimension = np.pad(shape[1:], (0, 1), constant_values=1)
    strides = np.cumprod(size_of_next_dimension[::-1], dtype='int32')[::-1]

    _, idx = jax.lax.scan(
        lambda remaining, stride: (
            remaining % stride,
            remaining // stride
        ),
        flat_index,
        strides
    )

    return tuple(idx)
```

Haven't looked at jax testing yet, would be helpful to know whether / how to incorporate these...

```python
def test_unravel_index():

    assert unravel_index(0, (2, 3)) == (0, 0)
    assert unravel_index(5, (2, 1, 3)) == (1, 0, 2)

    assert unravel_index(0, ()) == ()
    assert unravel_index(3, (2,)) == (3,)
    assert unravel_index(-1, (2, 1, 3)) == (-1, 0, 2)
```

### PR DESCRIPTION
**Pull request by mattwescott**
To maximize consistency with indexing a raveled array, negative indices are tolerated, even though `onp.unravel_index` would raise an exception. Out-of-bounds indices are clipped. 

Closes #2889

### CODE REVIEW

**Comment by mattwescott** [comment]
@googlebot signed

File: tests/lax_numpy_test.py [hunk 1]
```diff
@@ -1992,6 +1992,23 @@ def testRavel(self):
     args_maker = lambda: [rng.randn(3, 4).astype("float32")]
     self._CompileAndCheck(lambda x: x.ravel(), args_maker, check_dtypes=True)
 
+  @parameterized.parameters(
+    (0, (2, 1, 3)),
+    (5, (2, 1, 3)),
+    (0, ()))
```
Comment (jakevdp) [inline tests/lax_numpy_test.py:1998]: numpy's version also supports array inputs for the first entry; e.g.
```
>>> np.unravel_index([22, 41, 37], (7,6))
(array([3, 6, 6]), array([4, 5, 1]))
```
It would be useful to test this case as well.

File: jax/numpy/lax_numpy.py [hunk 2]
```diff
@@ -1153,6 +1153,17 @@ def ravel(a, order="C"):
   return reshape(a, (size(a),), order)
 
 
+@_wraps(onp.unravel_index)
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:1156]: Add a bit of documentation about the differences in behavior for negative and out-of-bounds indices (using the `lax_description` argument)

File: jax/numpy/lax_numpy.py [hunk 3]
```diff
@@ -1160,8 +1160,11 @@ def ravel(a, order="C"):
 
 @_wraps(onp.unravel_index, lax_description=_UNRAVEL_INDEX_DOC)
 def unravel_index(flat_index, shape):
+  flat_index = asarray(flat_index)
   sizes = pad(shape, (0, 1), constant_values=1)
   cumulative_sizes = cumprod(sizes[::-1])[::-1]
+  # Add enough trailing dims to avoid conflict with flat_index
+  cumulative_sizes = cumulative_sizes.reshape([-1] + [1] * flat_index.ndim)
```
Comment (mattwescott) [inline jax/numpy/lax_numpy.py:1167]: @jakevdp if you know a cleaner way to add trailing dim(s)...

Comment (jakevdp) [reply jax/numpy/lax_numpy.py:1167]: A useful pattern here would be to store the shape of the input, then `ravel()` it so the computation only has to handle 1D, then reshape the outputs before returning.

Comment (mattwescott) [reply jax/numpy/lax_numpy.py:1167]: Something like this?

```
idx = clipped_index.ravel() % cumulative_sizes[:-1] // cumulative_sizes[1:]
return tuple(idx.reshape([-1] + list(flat_index.shape)))
```

To my eye, the current implementation is slightly cleaner. Unless you are seeing other benefits, inclined to leave as is.

Comment (jakevdp) [reply jax/numpy/lax_numpy.py:1167]: No, I was saying that you ravel ``flat_index`` above so you can avoid the complicated reshape expression here, and then you return ``tuple([d.reshape(flat_index.shape) for d in idx])``

File: jax/numpy/lax_numpy.py [hunk 4]
```diff
@@ -1153,6 +1153,25 @@ def ravel(a, order="C"):
   return reshape(a, (size(a),), order)
 
 
+_UNRAVEL_INDEX_DOC = """\
+Unlike numpy's implementation of unravel_index, negative indices are accepted
+and out-of-bounds indices are clipped.
+"""
+
+@_wraps(onp.unravel_index, lax_description=_UNRAVEL_INDEX_DOC)
+def unravel_index(flat_index, shape):
```
Comment (jakevdp) [inline jax/numpy/lax_numpy.py:1162]: I just noticed: the numpy implementation names the first argument ``indices``; we should match that. It also accepts an ``order`` argument that defaults to ``C``: it would be useful to add that here and raise a NotImplementedError (or implement it, if you're up for that).

File: jax/numpy/lax_numpy.py (hunk 3 above)
Comment (mattwescott) [reply jax/numpy/lax_numpy.py:1167]: Sure could do multiple reshapes to avoid `[-1] + list(flat_index.shape)` . The current implementation still feels a bit cleaner to me.

File: jax/numpy/lax_numpy.py (hunk 4 above)
Comment (mattwescott) [reply jax/numpy/lax_numpy.py:1162]: Should this be a separate issue to detect mismatch and missing arguments in `_wraps`? Will let you take care of this in either case. Thanks for the reviews.

Comment (jakevdp) [reply jax/numpy/lax_numpy.py:1162]: Just to be clear: you're saying you'd like me to take over the PR to rename this variable name?

Comment (mattwescott) [reply jax/numpy/lax_numpy.py]: Added the rename and a clarifying change. Not sure I understand the barrier you are seeing to handling missing optional arguments in `_wraps`. Why would there be additional exceptions? Especially if you'll be updating the reference numpy version, might be less maintenance overall.

Comment (jakevdp) [reply jax/numpy/lax_numpy.py]: I tried quickly comparing wrapped and unwrapped function signatures via ``inspect.signature``, and a large number of functions differ (e.g. using args/kwargs vs explicit names, jax not implementing deprecated keywords, jax adding extra documented keywords, etc.) My feeling is there are too many exceptions to make a test like that useful.

**Review by jakevdp — APPROVED** [review]

File: jax/numpy/lax_numpy.py (hunk 4 above)
Comment (mattwescott) [reply jax/numpy/lax_numpy.py]: We must be looking at this differently, because such a degree of mismatch sounds like a case for checking optional arguments in `_wraps`. Moved to #2988
